components/tikv_util/src/trend.rs

```diff
@@ -0,0 +1,737 @@
+
+        // Spike entering check
+        if (value as f64) > enter_threshold {
+            // Hold the very high values in the checking sequence
+            self.values.push_back(curr);
+            if now.duration_since(self.values.front().unwrap().time) > self.duration {
+                // The checking sequence is too long to be a spike, dump all and exit checking
+                let values: Vec<SampleValue> = self.values.drain(..).collect();
+                return Some(values);
+            }
+            // The curr value is on hold, return None
+            return None;
+        }
+
+        // Not in a spike, nothing happen
+        if self.values.is_empty() {
+            return Some(vec![curr]);
+        }
+
+        // In a spike
+
+        // Spike ending check
+        if (value as f64) < exit_threshold {
+            if self.values.len() <= 2 {
+                // The checking sequence is too short to be a spike, dump all and exit checking
+                let mut values: Vec<SampleValue> = self.values.drain(..).collect();
+                values.push(curr);
+                return Some(values);
+            }
+            // The checking sequence is not long enough to be regular high, it's a spike,
+            // discard all but return curr
+            self.filter_value_gauge.set(self.avg() as i64);
+            self.filter_count_gauge.inc();
+            self.values.drain(..);
+            return Some(vec![curr]);
+        }
+
+        // Hold curr value to this spike
+        self.values.push_back(curr);
+        None
+    }
+
+    #[inline]
+    fn avg(&self) -> f64 {
+        if self.values.is_empty() {
+            return 0.0;
+        }
+        let mut sum: f64 = 0.0;
+        for value in self.values.iter() {
+            sum += value.value as f64;
+        }
+        sum / (self.values.len() as f64)
+    }
+}
+
+// Responsibilities of each window:
+//
+// L0:
+//     Eleminate very short time jitter,
+//     Consider its avg value as a point in data flow
+// L1:
+//     `L0.avg/L1.avg` to trigger slow-event, not last long but high sensitive
+//     Sensitive could be tuned by `L0.duration` and `L1.duration`
+//     Include periodic fluctuations, so it's avg could be seen as baseline
+// value     Its duration is also the no-detectable duration after TiKV starting
+// L2:
+//     `L1.avg/L2.avg` to trigger slow-event, last long but low sensitive
+//     Sensitive could be tuned by `L1.duration` and `L2.duration`
+//
+// L* History:
+//     Sample history values and calculate the margin error
+//
+// Spike Filter:
+//     Erase very high and short time spike-values
+//
+pub struct Trend {
+    sample_interval: usize,
+    sample_sequence_id: usize,
+
+    spike_filter: SpikeFilter,
+    spike_filter_enabled: bool,
+
+    data_flow: SampleWindows,
+
+    l1_history: HistoryWindow,
+    l2_history: HistoryWindow,
+
+    // When SPOT is being used, these should be `risk multiple`
```

Comment:
The SPOT algorithm have a param call `risk`, a very simple explanation(not fully right, but it's a simple saying) is:
* There is a data set, the highest/lowest `x%` we call it abnormal
* That `x%` is the `risk`, we pass it though a param to SPOT algorithm

Since we got more than one SPOT instance (if we use SPOT), then we will have more than one `risk` in all SPOT instances:
* We will give all SPOT instances a default `base risk`(different default values in different instances)
* Use param `risk multiple` to control the sensitiveness, each instance's `risk` will be `base risk` * `risk multiple`